src/main/java/io/micronaut/build/MicronautQualityChecksPlugin.java

```diff
@@ -0,0 +1,80 @@
+import org.gradle.api.plugins.BasePlugin;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.reporting.ReportingExtension;
+import org.gradle.testing.jacoco.plugins.JacocoCoverageReport;
+import org.gradle.testing.jacoco.plugins.JacocoPlugin;
+import org.gradle.testing.jacoco.plugins.JacocoReportAggregationPlugin;
+import org.gradle.testing.jacoco.tasks.JacocoReport;
+import org.sonarqube.gradle.SonarQubeExtension;
+import org.sonarqube.gradle.SonarQubePlugin;
+
+import java.util.Collections;
+
+@SuppressWarnings("UnstableApiUsage")
+public class MicronautQualityChecksPlugin implements Plugin<Project> {
+
+    public static final String COVERAGE_REPORT_TASK_NAME = "testCodeCoverageReport";
+    public static final String TEST_REPORT_TASK_NAME = "jacocoTestReport";
+
+    @Override
+    public void apply(final Project rootProject) {
+        if (!rootProject.equals(rootProject.getRootProject())) {
+            throw new GradleException("The Quality Checks plugin must only be applied to the root project");
+        }
+
+        configureSonar(rootProject);
+        configureJacoco(rootProject);
+        rootProject.getSubprojects().forEach(this::subprojectApply);
+    }
+
+    private void subprojectApply(final Project subproject) {
+        configureJacocoForSubproject(subproject);
+    }
+
+    private void configureSonar(final Project rootProject) {
+        rootProject.getPluginManager().apply(SonarQubePlugin.class);
+        final SonarQubeExtension sonarQubeExtension = rootProject.getExtensions().findByType(SonarQubeExtension.class);
+        final String githubSlug = (String) rootProject.findProperty("githubSlug");
+        sonarQubeExtension.properties(p -> {
+            p.property("sonar.projectKey", githubSlug.replaceAll("/", "_"));
+            p.property("sonar.organization", "micronaut-projects");
+            p.property("sonar.host.url", "https://sonarcloud.io");
+            p.property("sonar.java.source", "8");
+        });
+    }
+
+    private void configureJacoco(final Project rootProject) {
+        rootProject.getPluginManager().apply(BasePlugin.class);
+        rootProject.getPluginManager().apply(JacocoReportAggregationPlugin.class);
+
+        rootProject.getTasks().register(COVERAGE_REPORT_TASK_NAME, JacocoReport.class);
+
+        ReportingExtension reporting = rootProject.getExtensions().getByType(ReportingExtension.class);
+        reporting.getReports().register(COVERAGE_REPORT_TASK_NAME, JacocoCoverageReport.class,
+                r -> r.getTestType().set(TestSuiteType.UNIT_TEST));
+    }
+
+    private void configureJacocoForSubproject(Project subproject) {
+        if (subproject.getPlugins().hasPlugin(JavaPlugin.class)) {
+            subproject.getPluginManager().apply(JacocoPlugin.class);
+        }
+
+        // Do not generate reports for individual projects
+        if (!subproject.getTasks().matching(task -> task.getName().equals(TEST_REPORT_TASK_NAME)).isEmpty()) {
```

Comment:
:x: This breaks task configuration avoidance (it forces the configuration of all tasks even if you don't call them).
In practice, I don't think there's a good reason to disable any task. If the user wants to call them, so be it. If they don't, they won't be configured so it's a no-op.

Reply:
This was taken from the [official Gradle documentation](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/samples/incubating/java/jvm-multi-project-with-code-coverage-standalone/groovy/buildSrc/src/main/groovy/myproject.java-conventions.gradle#L23) 🤷 

I guess the point is to avoid report generation for each subproject, since individual reports will not be used at all.

Reply:
Well let's say I disagree with this recommendation, and that some Gradle APIs are obscure even to experts (`tasks.matching` is particularly nasty as it forces realization of tasks).